### PR TITLE
Refactor (de)convolution_2d

### PR DESCRIPTION
**Pull request by niboshi**
This is just a refactoring, in preparation for integrating iDeep.

* Moved `ndarray` type check to `function_node._check_data_type_forward`.
* Refactored duplicate code as a method (`_calc_out_size`)
* Isolated cuDNN forward calculation as methods.

### CODE REVIEW

File: chainer/functions/connection/convolution_2d.py [hunk 1]
```diff
@@ -216,6 +164,110 @@ def forward_gpu(self, inputs):
...
+            algo, workspace.data.ptr, workspace_size, zero.data,
+            y_desc.value, y.data.ptr)
+
+        # TODO(beam2d): Support unshared bias
+        if b is not None:
+            cudnn.add_tensor(
+                handle, one.data, bias_desc.value, b.data.ptr,
+                one.data, y_desc.value, y.data.ptr)
+
+        x = cuda.cupy.ascontiguousarray(x)
```
Comment (hvy) [inline chainer/functions/connection/convolution_2d.py]: The method body is duplicated. I suppose `y,` should be returned on this line.

**Comment by niboshi** [comment]
Thank you for finding a mistake!
I fixed the code.

**Comment by kmaehashi** [comment]
Jenkins test this please

**Comment by hvy** [comment]
It seems like the `NStep*` tests are failing on Jenkins on the new inputs type check in `FunctionNode.apply()`.

**Comment by hvy** [comment]
The failures above are caused by the indices being numpy arrays when calling permutate. https://github.com/chainer/chainer/blob/master/chainer/links/connection/n_step_gru.py#L128

**Comment by niboshi** [comment]
I removed ndarray type check from this PR and separated as #4029 and #4030.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
LGTM! Sorry for the delayed merge.

**Comment by hvy** [comment]
I'll create a backport after https://github.com/chainer/chainer/pull/4021 has been merged.

**Comment by kmaehashi** [comment]
Removed `to-be-backported` as it was not backported to v3.

**Pull request merged**